lib/ims/lti/models/messages/Message.rb

```diff
@@ -0,0 +1,57 @@
+require 'active_model'
+
+module Ims::Lti::Models::Messages
+  # Abstract base class for all messages.
+  #
+  # @attr [Hash] extensions Message parameters not defined in the specification.
+  # @attr [Hash] custom Parameters specified by the tool provider.
+  class Message
+    include ActiveModel::Model
+    include ActiveModel::Serialization
+    include ActiveModel::Validations
+
+    attr_accessor :extensions, :custom
+
+    def initialize(params = {})
+      singleton_class.class_eval do
+        attr_accessor *(required_parameters + recommended_parameters + optional_parameters)
```

Comment:
I wouldn't worry about classifying the parameters between required recommended optional. I would just define all of the IMS defined parameters as attr_accesors

Reply:
I would avoid modifying the singleton_class

Reply:
Sounds good. If we don't need to separate the param types I'll just go with a simpler approach I think.